test/Main.hs

```diff
@@ -1,107 +1,147 @@
-                      let tmp = "/tmp/test.lz4"
-                          openers = do
-                              r <- openFile f ReadMode
-                              w <- openFile tmp WriteMode
-                              return (r, w)
-                          closers (r, w) = hClose r >> hClose w
-                      (r, w) <- openers
-                      S.unfold H.readChunksWithBufferOf (300, r)
-                          & compress i
-                          & H.fromChunks w
-                      closers (r, w)
-                      f1 <-
-                          S.toList
-                              $ S.bracket_ (openFile tmp ReadMode) hClose
-                              $ \h ->
-                                    S.unfold H.readChunksWithBufferOf (50, h)
-                                        & decompress
-                                        & AS.concat
-                      f2 <-
-                          S.toList
-                              $ S.bracket_ (openFile f ReadMode) hClose
-                              $ \h -> S.unfold H.read h
-                      f1 `shouldBe` f2
+genArrayW8List :: Gen [Array.Array Word8]
+genArrayW8List = listOf $ Array.fromList <$> listOf (elements [0,1])
+
+genArrayW8Large :: Gen (Array.Array Word8)
+genArrayW8Large = do
+    let minArr = 1024 * 10
+        maxArr = 1024 * 100
+    arrS <- choose (minArr, maxArr)
+    Array.fromList <$> vectorOf arrS (elements [0,1])
+
+genArrayW8ListLarge :: Gen [Array.Array Word8]
+genArrayW8ListLarge = do
+    sz <- choose (50, 100)
+    vectorOf sz genArrayW8Large
+
+genAcceleration :: Gen Int
+genAcceleration = elements [-1..12]
+
+decompressCompressChunk :: Int -> Array.Array Word8 -> IO ()
+decompressCompressChunk i arr = do
+    lz4Ctx <- c_createStream
+    lz4CtxD <- c_createStreamDecode
+    compressed <- compressChunk i lz4Ctx arr
+    decompressed <- decompressChunk lz4CtxD compressed
+    c_freeStream lz4Ctx
+    c_freeStreamDecode lz4CtxD
+    decompressed `shouldBe` arr
+
+decompressCompressChunk2 ::
+       Int -> Array.Array Word8 -> Array.Array Word8 -> IO ()
+decompressCompressChunk2 i arr1 arr2 = do
+    lz4Ctx <- c_createStream
+    lz4CtxD <- c_createStreamDecode
+    compressed1 <- compressChunk i lz4Ctx arr1
+    compressed2 <- compressChunk i lz4Ctx arr2
+    decompressed1 <- decompressChunk lz4CtxD compressed1
+    decompressed2 <- decompressChunk lz4CtxD compressed2
+    c_freeStream lz4Ctx
+    c_freeStreamDecode lz4CtxD
+    (decompressed1, decompressed2) `shouldBe` (arr1, arr2)
+
+decompressResizedcompress :: Int -> [Array.Array Word8] -> IO ()
+decompressResizedcompress i lst =
+    let strm = Stream.fromList lst
+     in do lst1 <- Stream.toList $ decompressResized $ compress i strm
+           lst `shouldBe` lst1
+
+decompressCompress :: Int -> Int -> [Array.Array Word8] -> IO ()
+decompressCompress bufsize i lst = do
+    let strm = Stream.fromList lst
+    tmp <- emptySystemTempFile "LZ4"
```

Comment:
We are leaving the temp file around when the test is done. We should use `withSystemTemFile` instead.